### DISCUSSION TITLE
Is there any way to implement screenshots in combination with Junit for GitLab?

### DISCUSSION THREAD

**Discussion by GhostLyrics** [jest-community/jest-junit#208] [Q&A] [answered]
I was checking out how to use [GitLab's support for screenshots as part of Junit reports](https://docs.gitlab.com/ee/ci/unit_test_reports.html#viewing-junit-screenshots-on-gitlab) which requires one to implement roughly the following:

```XML
<testcase time="1.00" name="Test">
  <system-out>[[ATTACHMENT|/path/to/some/file]]</system-out>
</testcase>
```

I saw [a post from 2019 in the closed issues](https://github.com/jest-community/jest-junit/issues/99) and was wondering whether anything had changed from then and it's now possible to implement this. If it's not and contributions are acceptable, maybe you could give some pointers where to start?

**Comment by palmerj3** [comment, marked as answer]
Thanks for submitting an issue!

If the syntax requires the use of system-out then that raises a few issues:

1. Jest reporters [don't have reliable access to anything sent to stdout/stderr](https://github.com/jest-community/jest-junit/pull/96#issuecomment-700168772) during a test run so you couldn't simply have jest output that attachment syntax and expect jest-junit to pick it up and create that system-out tag.
2. Right now we only offer a way to add testsuite properties dynamically, but not make changes to testcase or add other tags.

So the only way I can think to implement this would be to use jest-junit as-is and then have a posttest script that looks for screenshots and modifies junit.xml to include the tags.

Or if you wanted jest-junit to do it then we'd either need a property specifically for gitlab which tells it to look for screenshots here, and if one is found, add the tag. Or have some mechanism whereby we can call out to a script users of jest-junit would maintain that would be responsible for making changes to a given testcase object before it's transformed to XML.

Hope this helps

**Reply by CMCDragonkai** [reply]
> Jest reporters https://github.com/jest-community/jest-junit/pull/96#issuecomment-700168772 during a test run so you couldn't simply have jest output that attachment syntax and expect jest-junit to pick it up and create that system-out tag.

Is there any workaround to get the STDOUT and STDERR produced during a test into the jest junit report? We have a lot of tests that may require load balancing between CI jobs, and some of our tests produce STDOUT and STDERR output, and these are essential for debugging why a test fails.

**Comment by palmerj3** [comment]
For a while now Jest has not passed along stdout/stderr to reporters from test runs. So a change would need to be made to jest itself so that reporters have this information like we used to.

**Reply by CMCDragonkai** [reply]
I can control how my loggers are logging out. Right now they are going directly to stderr. Is there a way to funnel them to instead jest so that they show up in the reports? I'm guessing they override/monkeypatch the console.error?

**Reply by CMCDragonkai** [reply]
One issue is that jest seems to buffer up these console.log messages and I'd like to see them unbuffered when debugging tests.

**Reply by palmerj3** [reply]
The output is buffered because jest runs test suites in parallel. Unbuffered would be completely unreadable unless you ran it with the runInBand flag which tells jest to only use one worker for test runs.

**Reply by CMCDragonkai** [reply]
I see, so right now my CI/CD runners are all single-process anyway, so perhaps I can just run them all `--runInBand`.

At the same time, redirecting my loggers to send messages to `console.error` instead (which is intercepted by jest).

That should enable my reporters to capture the relevant logs for each test run.
